scanners/seo.py

```diff
@@ -0,0 +1,146 @@
+    logging.debug("Scan function called with options: %s" % options)
+
+    # Initialize results with a quick check for the presence of robots.txt.
+    # Because we're not reading robots.txt, we only need the status code.
+    # We could put these in `pages` but since we have specialized needs for each
+    # and no need to iterate through them, I don't want to.
+    results = {
+        'robots': requests.get("https://" + domain + '/robots.txt').status_code,
+        'warnings': {}
+    }
+    # Now let's take a look at sitemap
+    sitemap = requests.get("https://" + domain + '/sitemap.xml')
+    sitemap_status = sitemap.status_code
+    results['sitemap'] = {
+        'status': sitemap_status
+    }
+    # If we have found a sitemap, see how many URLs are in there?
+    # TO DO: If sitemap.xml redirects to a sitemap index,
+    # we're not handling that correctly yet. We need to detect that
+    # redirect, then ... what? Follow from there?
+    if sitemap_status == 200:
+        soup = BeautifulSoup(sitemap.text, 'xml')
+        urls = soup.find_all('url')
+        results['sitemap']['urls found'] = len(urls)
+        # and how many of those URLs appear to be PDFs
+        if len(urls) > 0:
+            pdfcount = len([u for u in urls if '.pdf' in u.get_text()])
+            results['sitemap']['PDFs found in sitemap'] = pdfcount
+
+    # We'll write to these empty lists for simple dupe checking later
+    titles = []
+    descriptions = []
+    # Perform the "task".
+    for page in environment['pages']:
+        try:
+            r = requests.get("https://" + domain + page, allow_redirects=True, timeout=4)
+            htmlsoup = BeautifulSoup(r.text, 'lxml')
+            # get title and put in dupe-checking list
+            title = htmlsoup.find('title').get_text()
+            titles.append(title)
+            # and description
+            description = htmlsoup.select_one("meta[name='description']")
+            if description:
+                descriptions.append(description['content'])
+            # and can we find dc:date?
+            dc_date = htmlsoup.select_one("meta[name='DC.Date']")
+            # if we found one, grab the content
+            if dc_date:
+                dc_date = dc_date['content']
+
+            # Find the main tag (or alternate)
+            # Potential TO-DO: check that there is only one. Necessary? ¯\_(ツ)_/¯
+            maintag = True if htmlsoup.find('main') else False
+            # if we couldn't find `main` look for the corresponding role
+            if not maintag:
+                maintag = True if htmlsoup.select('[role=main]') else False
+
+            results[page] = {
+                'page': page,
+                'status': str(r.status_code),
+                'title': title,
+                'description': description,
+                'has main tag': maintag,
+                'DC Date': dc_date
+            }
+        except Exception as error:
```

Comment:
I'm not totally sure, but it looks like this `except` block is handling network errors from `requests`? Thoughts on moving it up closer to the `requests.get` call? Then we can handle the different network errors as necessary.

```
try:
    r = requests.get(url, params={'s': thing})
except requests.exceptions.Timeout:
    # Maybe set up for a retry, or continue in a retry loop
except requests.exceptions.TooManyRedirects:
    # Tell the user their URL was bad and try a different one
except requests.exceptions.RequestException as e:
    # catastrophic error. bail.
    raise SystemExit(e)
```

^ From [this SO comment](https://stackoverflow.com/a/16511493).

Reply:
It's really capturing any error in getting the various data. That's why it's writing the error to results. So you can hopefully figure out what happened. In testing, most of the errors were because the page wasn't as expected -- which should be expected.

Reply:
Gotcha. And that error data would be visible in the CSV/JSON?

Reply:
Should be, yeah. It's in results, which is cast to csv.

Reply:
Hold the phone. Only part of results is NOT getting to CSV. Stand by while I figure out why the rest is getting thrown out.